src/sources/datadog_agent/logs.rs

```diff
@@ -192,3 +201,89 @@ pub(crate) fn decode_log_body(
 
     Ok(decoded)
 }
+
+// ddtags input is a string containing a list of key-value pairs.
+// the tag-value list members are separated by `,` and the
+// tag-value pairs are separated by `:`.
+// The output is an Object constructed with the tag-value pairs.
+//
+// Technically, the Datadog logs intake accepts input where this
+// field is a bare tag (no `:`), so we're supporting that here,
+// in which case the raw input is not expanded into an object.
+fn parse_ddtags(ddtags_raw: &Bytes) -> Value {
+    let ddtags_str = String::from_utf8_lossy(ddtags_raw);
+
+    // The value is either an empty string or a bare tag
+    if !ddtags_str.contains(',') && !ddtags_str.contains(':') {
+        return Value::Bytes(ddtags_raw.clone());
+    }
+
+    // There is at least one key-value pair
+    let ddtags_object: ObjectMap = ddtags_str
+        .split(',')
+        .filter(|kv| !kv.is_empty())
+        .map(|kv| match kv.split_once(':') {
+            Some((k, v)) => (KeyString::from(k), Value::Bytes(Bytes::from(v.to_string()))),
+            None => (KeyString::from(kv), Value::Null),
+        })
+        .collect();
+
+    if ddtags_object.is_empty() && !ddtags_str.is_empty() {
+        warn!(message = "`parse_ddtags` set to true and Agent log contains non-empty ddtags string, but no tag-value pairs were parsed.")
+    }
+
+    ddtags_object.into()
+}
+
+#[cfg(test)]
+mod tests {
+    use super::*;
+
+    #[test]
+    fn ddtags_parse_empty() {
+        let raw = Bytes::from(String::from(""));
+        let val = parse_ddtags(&raw);
+
+        assert!(val == Value::Bytes(Bytes::from("")));
```

Comment:
These should be `assert_eq!`, and possibly add `use similar_asserts::assert_eq;` to get better comparisons for failures.
```suggestion
        assert_eq!(val, Value::Bytes(Bytes::from("")));
```

Reply:
> add use similar_asserts::assert_eq; to get better comparisons for failures.

😱  this is amazing. Where has this been all my life

Reply:
There's also [pretty-assertions](https://docs.rs/pretty_assertions/latest/pretty_assertions/) but we seem to have settled on `similar-asserts`.